src/common/json.cc

```diff
@@ -0,0 +1,608 @@
+/*!
+ * Copyright (c) by Contributors 2019
+ */
+#include <sstream>
+
+#include "xgboost/logging.h"
+#include "xgboost/json.h"
+#include "xgboost/json_io.h"
+#include "../common/timer.h"
+
+namespace xgboost {
+
+void JsonWriter::Save(Json json) {
+  json.ptr_->Save(this);
+}
+
+void JsonWriter::Visit(JsonArray const* arr) {
+  this->Write("[");
+  auto const& vec = arr->getArray();
+  size_t size = vec.size();
+  for (size_t i = 0; i < size; ++i) {
+    auto& value = vec[i];
+    this->Save(value);
+    if (i != size-1) { Write(", "); }
+  }
+  this->Write("]");
+}
+
+void JsonWriter::Visit(JsonObject const* obj) {
+  this->Write("{");
+  this->BeginIndent();
+  this->NewLine();
+
+  size_t i = 0;
+  size_t size = obj->getObject().size();
+
+  for (auto& value : obj->getObject()) {
+    this->Write("\"" + value.first + "\": ");
+    this->Save(value.second);
+
+    if (i != size-1) {
+      this->Write(",");
+      this->NewLine();
+    }
+    i++;
+  }
+  this->EndIndent();
+  this->NewLine();
+  this->Write("}");
+}
+
+void JsonWriter::Visit(JsonNumber const* num) {
+  convertor_ << num->getNumber();
+  auto const& str = convertor_.str();
+  this->Write(StringView{str.c_str(), str.size()});
+  convertor_.str("");
+}
+
+void JsonWriter::Visit(JsonRaw const* raw) {
+  auto const& str = raw->getRaw();
+  this->Write(str);
+}
+
+void JsonWriter::Visit(JsonNull const* null) {}
+
+void JsonWriter::Visit(JsonString const* str) {
+  std::string buffer;
+  buffer += '"';
+  auto const& string = str->getString();
+  for (size_t i = 0; i < string.length(); i++) {
+    const char ch = string[i];
+    if (ch == '\\') {
+      if (i < string.size() && string[i+1] == 'u') {
+        buffer += "\\";
+      } else {
+        buffer += "\\\\";
+      }
+    } else if (ch == '"') {
+      buffer += "\\\"";
+    } else if (ch == '\b') {
+      buffer += "\\b";
+    } else if (ch == '\f') {
+      buffer += "\\f";
+    } else if (ch == '\n') {
+      buffer += "\\n";
+    } else if (ch == '\r') {
+      buffer += "\\r";
+    } else if (ch == '\t') {
+      buffer += "\\t";
+    } else if (static_cast<uint8_t>(ch) <= 0x1f) {
+      // Unit separator
+      char buf[8];
+      snprintf(buf, sizeof buf, "\\u%04x", ch);
+      buffer += buf;
+    } else {
```

Comment:
I would suggest to have another branch static_cast<uint8_t>(ch) > 127 in which we would fail serialization. Supporting UTF8 encoded Unicode strings is hard and I suggest at least for now limit ourself to the realm of simple strings